patches/0004-Output-to-hashed-mappings.patch

```diff
@@ -0,0 +1,1622 @@
++		public static final String OUT_MAP_FILE_NAMED_INTERMEDIARY = "outMapFileNamedHashed";
+ 		public static final String OUT_REFMAP_FILE = "outRefMapFile";
+ 		public static final String DEFAULT_OBFUSCATION_ENV = "defaultObfuscationEnv";
+ 		public static final String QUIET = "quiet";
+diff --git a/src/main/java/net/fabricmc/loom/util/ModRemappingHelper.java b/src/main/java/net/fabricmc/loom/util/ModRemappingHelper.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..839777b63d2f06c257f09e9fcba7088ed89972e8
+--- /dev/null
++++ b/src/main/java/net/fabricmc/loom/util/ModRemappingHelper.java
+@@ -0,0 +1,57 @@
++/*
++ * This file is part of fabric-loom, licensed under the MIT License (MIT).
++ *
++ * Copyright (c) 2022 FabricMC
++ *
++ * Permission is hereby granted, free of charge, to any person obtaining a copy
++ * of this software and associated documentation files (the "Software"), to deal
++ * in the Software without restriction, including without limitation the rights
++ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
++ * copies of the Software, and to permit persons to whom the Software is
++ * furnished to do so, subject to the following conditions:
++ *
++ * The above copyright notice and this permission notice shall be included in all
++ * copies or substantial portions of the Software.
++ *
++ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
++ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
++ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
++ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
++ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
++ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
++ * SOFTWARE.
++ */
++
++package net.fabricmc.loom.util;
++
++import java.io.ByteArrayInputStream;
++import java.io.File;
++import java.io.IOException;
++import java.io.InputStream;
++import java.io.InputStreamReader;
++
++import com.google.gson.JsonObject;
++
++import net.fabricmc.loom.LoomGradlePlugin;
++import net.fabricmc.loom.api.mappings.layered.MappingsNamespace;
++
++public final class ModRemappingHelper {
++	public static MappingsNamespace getIntermediateMappings(File source) {
++		if (!ZipUtils.contains(source.toPath(), "quilt.mod.json")) return MappingsNamespace.INTERMEDIARY;
++
++		InputStream qmj;
++
++		try {
++			qmj = new ByteArrayInputStream(ZipUtils.unpack(source.toPath(), "quilt.mod.json"));
++		} catch (IOException e) {
++			throw new RuntimeException(e);
++		}
++
++		JsonObject jsonObject = LoomGradlePlugin.GSON.fromJson(new InputStreamReader(qmj), JsonObject.class);
++		String mappings = jsonObject.getAsJsonObject("quilt_loader").get("intermediate_mappings").getAsString();
++
++		if (mappings.equals("net.fabricmc:intermediary")) return MappingsNamespace.INTERMEDIARY;
++
++		return MappingsNamespace.HASHED;
```

Comment:
This logic is unsound; it will assume *any* "intermediate_mappings" value besides fabric intermediary is hashed, but with loader plugins we can't promise that's the case

Reply:
That's true, I did this so it would fall back to hashed. I can instead have it explicitly check for hashed or an empty value (implicitly hashed) and have it throw an exception if the value's unknown if that's preferred.